### ISSUE TITLE
Misalignment of Header when scrollX set to true

### ISSUE THREAD

**Issue by csturm83** [DataTables/DataTables#539]
This may be similar to <a href="https://github.com/DataTables/DataTables/issues/486" target="_blank">Issue #486</a>.

When scrollX is set to true and the content doesn't require a scroll bar, the header table appears to be aligned to the left and the body table appears aligned to center. Here is a <a href="http://jsfiddle.net/csturm83/wrugvxst/" target="_blank">jsfiddle</a> demonstrating this. You may need to drag the vertical divider in jsfiddle to the left to see it pronounced.

It is my understanding that both tables have the class <b>dataTable</b> with <b>margin: 0 auto;</b>.

I noticed that when scrollX is enabled the header table is wrapped in a div with class <b>dataTables_scrollHeadInner</b> which throws off alignment slightly.

My workaround was to add the following css to keep the header centered along with the body:

.dataTables_scrollHeadInner {
margin-right: auto;
margin-left: auto;
}

**Comment by DataTables** [comment]
Hi,

Thanks for the JSFiddle. The problem was a bug in 1.10.5, but this was fixed in the 1.10.6 release. I've updated your fiddle to use the latest version here: http://jsfiddle.net/wrugvxst/13/ .

Allan

**Comment by csturm83** [comment]
I am still able to reproduce this issue (slightly less trivially) in the following <a href="http://jsfiddle.net/4cz98p0x/" target="_blank">jsfiddle</a> using 1.10.6. Again, you may need to drag the vertical splitter.

**Comment by DataTables** [comment]
With the `autoWidth` option set to false, DataTables will not automatically calculate column widths and adjust them as the page resizes. You need to leave this set to `true` (which it is by default).

**Comment by csturm83** [comment]
Thanks for the clarification. I will admit I am relatively new to DataTables and was somewhat confused as to why the autoWidth option would affect the alignment of the columns as well. For my particular use case, changing this option is not an issue. I just wonder if it is safe to assume that there is not a valid use case for that particular configuration.

Also, the jsfiddle I included is greatly simplified. I didn't actually notice the issue while resizing the window. I just zeroed in on the configuration that seemed to cause the issue, and that's how it best presents itself.

The context of where I see the issue is a little more involved. I currently have a DataTable with data that has some data that would require scrolling, while some of it does not. The issue appears when I have all of the records up front (which requires scrolling) and then enter a search term that zeros in on data that doesn't require scrolling. This in and of itself doesn't reproduce the issue. I also have stateSave enabled so that the search term is persisted. One of the columns contains hyperlinks (using mRender) which navigates to an update page. The issue is presented when I click the link on the filtered data and navigate back to the page with the DataTable and persisted search term (via another hyperlink). Again, changing autoWidth to true corrects this issue. I just wanted to point out that it's not limited to page resizing - which you might already know. Just thought I'd mention it.

Anyways, thanks for your help on this. I apologize if not all of the information I have provided is useful or even necessary.

**Comment by DataTables** [comment]
To a large extend I regret putting the autoWidth option in as disabling it has all sorts of unpleasant side effects. I need to put a big "WARNING" label in the documentation!

Thanks for the additional information that is useful! I will take a look at the implementation to remind myself why this problem occurs when autoWidth is disabled (can't remember off the top of my head!), but I don't think it is likely to change in the near future.

**Comment by gregmarr** [comment]
I'm seeing this in 1.10.12, and I haven't touched autoWidth.

    <!DOCTYPE html>
    <html>
        <head>
            <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script>
            <link rel="stylesheet" type="text/css" href="https://cdn.datatables.net/1.10.12/css/jquery.dataTables.css">
            <script type="text/javascript" charset="utf8" src="https://cdn.datatables.net/1.10.12/js/jquery.dataTables.js"></script>
            <script>$(document).ready( function () {
                $('#results').DataTable({ scrollX: true });
            } );</script>
            <style>
                table { width:100%; }
                td { border:darkgray 1px solid; }
            </style>
        </head>
        <body>
            <table id="results">
                <thead>
                <tr>
                    <th>Workspace Name</th>
                    <th>Mean Elapsed</th>
                    <th>Last Executed</th>
                    <th>Status Code</th>
                    <th>Response</th>
                    <th>Data</th>
                </tr>
                </thead>
                <tbody>
                    <tr>
                        <td>unicodeInIronpython</td>
                        <td class=failure>0s</td>
                        <td>0001-01-01 00:00:00 &#43;0000 UTC</td>
                        <td>0</td><td></td><td></td>
                    </tr>
                </tbody>
            </table>
        </body>
    </html>

<img width="1349" alt="screen shot 2016-11-29 at 5 30 53 pm" src="https://cloud.githubusercontent.com/assets/8569738/20731756/b049348e-b659-11e6-855b-d1532485b2ed.png">

This is Chrome on Mac.  The misalignment happens when the width of the content is less than the width of the browser.

**Comment by gregmarr** [comment]
My issue looks to be https://github.com/DataTables/DataTables/issues/794

**Comment by DataTables** [comment]
@gregmarr - Add `width="100%"` to your table tag and it will work okay.

**Comment by ZZa** [comment]
The problem reproduces today, in version **1.10.16**.

Adding the property `width="100%"` to the target table fixes problem with small number of columns (when the screen is bigger than the table), but ruins logic of width property for column when there're plenty of columns. 

My workaround was to add `min-width: 100%` style to my table, but it feels like an ad-hoc. Still needs to be reviewed.

**Comment by ndmeiri** [comment]
I can also reproduce the problem in v1.10.16. At least on Safari on a Mac. The above workaround of adding `min-width: 100%` to my table also seems to work. But can we please reopen this issue? It would be nice if the problem was removed from the library instead of requiring a workaround.

**Comment by brian-lamb-software-engineer** [comment]
I managed to fix the issue that i was having, where when i set `"scrollX": true,` the header would not display, like this
`.dataTables_sizing{
   height: 30px !important;
 }`

**Comment by ocervell** [comment]
If you're using semantic UI, you can forget about "scrollX" option.
Instead, make the table wrapper a 'ui container' and set 'overflow-x': 'auto':

$("#table-id").parent().addClass('ui container special')
.ui.container.special {
  'overflow-x': 'auto';
}

**Comment by kjkrum** [comment]
`.dataTables_scrollHead` or `.dataTables_scrollHeadInner` needs right padding when `.dataTables_scrollBody` has a vertical scroll bar.

**Comment by pierbuca** [comment]
I have similar problem, but I can reproduce the issue only on chrome (last version at time of writing).
I tried workaround suggested above but without any success. 
I have a table with a large number of columns and this intialition parameter:
`select: true,
"sScrollY": "400px",
"scrollCollapse": true,
 "paging": false,`

I'm using 1.10.16 version.
Any news on this issue relolution?

**Comment by jschortz2** [comment]
I just started working with dataTables yesterday, and immediately encountered the header misalignment when using scrollY.  I had included the FixedHeader extension, but after tinkering with it for a few hours, discovered that eliminating that extension entirely, while still using the fixedHeader initialization option, allowed my test script to work flawlessly - even with multi-row/multi-column spanning headers.  See standalone test attached.

What I'm wondering is, has the FixedHeader functionality become part of dataTables core, and if so, why is the incompatible extension still part of the download page?  Very, very happy with dataTables, though, despite the hours of initial frustration :-) 
[testdt1.zip](https://github.com/DataTables/DataTables/files/1691835/testdt1.zip)

**Comment by DataTables** [comment]
FixedHeader is an extension to DataTables and will not become part of the core. The built in scrolling options are not the same as FixedHeader - although they overlap to some extend. It is also not compatible with the scrolling options as [noted in the [compatibility chart](https://datatables.net/download/compatibility).

It should really throw an error when enabled with the built in scrolling options.

**Comment by mseleng** [comment]
Hey, guys, I am still having an issue even though I set `width:100%`. I got 9 columns in my table and the header is misaligned only on page load. If I resize the page, the header seems to realign itself correctly. I am using only horizontal scroll, `scrollX: true`. Any ideas?

**Comment by DataTables** [comment]
You probably need to call `columns.adjust()` once everything has loaded so DataTables can take into account the width of the images / icons when drawn. Either that or specify the width as an absolute measurement in your CSS.

**Comment by mseleng** [comment]
Yeah, that's one possibility, but in my case, I found it much simple to just plain css fontawesome :)

**Comment by alexscg** [comment]
I also was just fighting with this, I'm on win/chrome... after reading all the comments and trying all the solutions what worked for me is to simply call resize on the window:  $(window).resize();

**Comment by wolfram77** [comment]
I saw it got fixed when i clicked the header or resized window. After resizing window (with settimeout 0) it is working fine now. I saw your comment later, but thanks.
